### PR TITLE
assign dl24-nezuko: vol-loss-weight-2.0-no-gradnorm

### PR DESCRIPTION
**Pull request by morganmcg1**
# Hypothesis: Static vol-loss-weight=2.0 WITHOUT GradNorm (Removes Self-Cancellation)

## Background

PR #911 definitively confirmed that combining `volume-loss-weight=2.0` with GradNorm is **self-cancelling**: GradNorm's adaptive mechanism drives task loss weights back toward equilibrium, neutralising the static upweighting within ~3 epochs. The static factor adds gradient noise without a persistent vol_p upweighting benefit.

The central unsolved problem in this wave is the systematic **val→test vol_p gap (+7pp)** confirmed across every completed long run:
- PR #800: val=4.28% → test=12.04% (+7.76pp)
- PR #806: val=4.07% → test=12.03% (+7.96pp)
- PR #844: val=4.06% → test=11.16% (+7.10pp)
- PR #874: val=4.13% → test=11.36% (+7.23pp)
- PR #923: val=4.80% → test=12.03% (+7.23pp, EMA variant)

Issue #882 mandates **reducing TEST vol_p** as the primary research objective. To close the val→test gap, we need the model to pay more attention to volume pressure *during training*, not just during evaluation.

## Hypothesis

**Static `volume-loss-weight=2.0` without GradNorm provides a persistent, non-self-cancelling upweighting of vol_p gradient signal during training.** Without GradNorm, the 2.0 weight is not equilibrated away — it remains active for all 50 epochs and forces the model to prioritise vol_p at the cost of potentially slightly increasing other component losses.

This is mechanistically distinct from PR #911 (which failed due to GradNorm cancellation), and directly tests whether persistent vol_p upweighting can compress the val→test gap.

## Configuration

Base: frieren #914 / tanjiro #900 validated triple-stack (5L STRING 5-oct + Y-sym p=0.5 + WD=0.005) plus:
- `--no-use-gradnorm` (disables GradNorm)
- `--volume-loss-weight 2.0` (persistent 2× upweight on vol pressure loss)
- `--train-volume-points 65000 --eval-volume-points 65000` (standard 65k)

## Current Baseline (in-wave val leaders, as of 2026-05-10)

| PR | Run | Student | EP | val_abupt | vol_p |
|---|---|---|---:|---:|---:|
| #914 | `wdxtdmhy` | dl24-frieren | EP20 | **6.6070%** | 4.1830% |
| #900 | `os6v64lq` | dl24-tanjiro | EP25 | **6.6160%** | 4.2508% |

Wave SOTA test: PR #740 `5x8wofzm` test_abupt=**7.5195%**, vol_p=10.758%

**Target to beat for merge consideration: test_primary/abupt_axis_mean_rel_l2_pct ≤ 7.5195%**

## Train Command

```bash
torchrun --nproc_per_node=8 train.py \
  --lr 1e-4 --weight-decay 0.005 --optimizer lion \
  --model-layers 5 --model-pe string_multisigma \
  --pe-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --no-use-gradnorm \
  --volume-loss-weight 2.0 \
  --use-y-symmetry-aug --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65000 --eval-volume-points 65000 \
  --epochs 50 \
  --wandb-group vol-loss-weight-static \
  --wandb-name nezuko/vol-loss-weight-2.0-no-gradnorm \
  --agent nezuko
```

## Gate Schedule (kill on miss)

| Gate | Epoch | Threshold | Action if miss |
|------|------:|----------:|---------------|
| EP5  | 5  | ≤ 7.5%  | KILL |
| EP10 | 10 | ≤ 7.2%  | KILL |
| EP15 | 15 | ≤ 6.80% | KILL |
| EP20 | 20 | ≤ 6.70% | KILL |
| EP25 | 25 | ≤ 6.65% | KILL |
| EP30 | 30 | ≤ 6.60% | KILL (terminal gate — close PR if final) |

**Important:** Standard epoch size (5,494 steps/epoch on DDP8 at batch_size=1) applies. Report metrics at each epoch boundary.

## Metrics to Report

At each gate epoch, please report:
- `val_primary/abupt_axis_mean_rel_l2_pct` (primary gate metric)
- `val_primary/volume_pressure_rel_l2_pct` (vol_p — key focus metric)
- `val_primary/surface_pressure_rel_l2_pct` (surf_p)
- `val_primary/wall_shear_rel_l2_pct` (wall)
- W&B run ID

**Critical:** Also report the raw loss weights at gate checkpoints to confirm vol_loss_weight is not being equilibrated (should remain at 2.0 throughout).

## Suggested Follow-Up

If vol_p shows improvement (val vol_p < 4.0%) but abupt is lagging, consider vol-loss-weight=1.5 in next round (less aggressive upweighting). If vol_p shows no improvement over baseline (~4.2%), the hypothesis is rejected and data-distribution approaches (OOD test-set alignment) should be explored.

## Expected Outcome

- **Success signal**: val vol_p drops below 4.0% (below #914's 4.183%) with abupt tracking gates
- **Failure signal**: abupt misses EP10 gate (≤7.2%) or vol_p shows no improvement over baseline
- **Neutral**: abupt tracks gates but vol_p unchanged (would suggest vol_p gap is not a training-time signal problem)

---

_Assigned by advisor on 2026-05-10. Issue #882 directive: prioritise TEST vol_p reduction._

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Your GPUs are running at 90-98% utilization which indicates training is underway, but you have not posted any launch confirmation or progress updates to this PR. Please immediately post a comment confirming: (1) the W&B run ID, (2) the current epoch and abupt metric, (3) gate status. We need this information to monitor the experiment. If you are having any issues posting updates, let me know.

**Comment by morganmcg1** [comment]
STUDENT nezuko: Apologies for the missing launch confirmation, advisor.

## Launch Confirmation

- **W&B run ID:** `6gd9u34e` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/6gd9u34e))
- **Launched:** 2026-05-10T00:53:18Z (~32 min ago)
- **Current state:** Epoch 3/50 ~55% complete (~6.65 it/s, all 8 GPUs at 92-98% util)

## Config Verification (from W&B)

| Key | Value |
|---|---|
| `volume_loss_weight` | **2** |
| `use_gradnorm` | **False** ✓ |
| `lr` | 1e-4 |
| `optimizer` | lion |
| `weight_decay` | 0.005 |
| `model_layers` | 5 |
| `model_pe` | string_multisigma |
| `pe_init_sigmas` | 0.25,0.5,1.0,2.0,4.0 |
| `use_y_symmetry_aug` | True (p=0.5) |
| `train_volume_points` / `eval_volume_points` | 65000 / 65000 |
| `epochs` | 50 |

Triple-stack verified, GradNorm disabled, static vol-weight=2.0 confirmed.

## Early Validation Metrics

| EP | abupt | vol_p | surf_p | wall_shear |
|---:|---:|---:|---:|---:|
| 1 | 17.878% | 9.991% | 12.927% | 19.926% |
| 2 | 11.858% | 6.805% | 8.445% | 13.226% |

Strong descent: abupt dropped 6.02pp from EP1→EP2. Trajectory is on track for the EP5 gate (≤ 7.5%).

## Gate Status

| Gate | Status |
|------|--------|
| EP5 ≤ 7.5%   | pending (next checkpoint, ~30 min) |
| EP10 ≤ 7.2%  | pending |
| EP15 ≤ 6.80% | pending |
| EP20 ≤ 6.70% | pending |
| EP25 ≤ 6.65% | pending |
| EP30 ≤ 6.60% | pending |

## Note on Launch Issue

The runs/ directory has a stale `nezuko_volw2_nogn.log` from a duplicate launch attempt at 01:02 that failed cleanly with `EADDRINUSE` (the original run was already holding port 29500). The original run at 00:53 is healthy and unaffected. Will report at EP5.

**Comment by morganmcg1** [comment]
**ADVISOR — Launch Confirmed, EP5 Gate Upcoming**

Thanks for the launch confirmation and EP2 metrics. Configuration looks correct:
- Static vol-loss-weight=2.0 ✓
- GradNorm disabled ✓
- 5-layer STRING 5-oct PE ✓
- Y-symmetry augmentation ✓

EP2 trajectory at 11.858% looks healthy — the -1.097%/1k step slope is strong descent. I'm monitoring W&B directly. At this slope, EP5 projection is approximately 7.1–7.5%, which is right at the gate.

**EP5 gate threshold: ≤7.5%** — please post as soon as EP5 validation completes with the full table.

The key question is whether the static vol-weight=2.0 (without GradNorm cancellation) produces a better vol_p trajectory than the reference frieren/tanjiro runs. At EP2 you're at vol_p=6.8% — frieren/tanjiro were at ~10% vol_p at EP2, so vol_p is already running hotter, consistent with the hypothesis that vol-loss upweighting is taking effect.

**Comment by morganmcg1** [comment]
STUDENT nezuko:
SENPAI-RESULT: {"terminal":true,"status":"killed_gate_miss","pending_arms":false,"wandb_run_ids":["6gd9u34e"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":9.0101},"test_metric":null}

## Results — EP5 GATE MISS — KILLED

The static `volume-loss-weight=2.0` without GradNorm produced the **opposite** of the hypothesised effect: vol_p worsens and overall convergence slows substantially. EP5 abupt missed the ≤7.5% gate by **+1.51pp** and vol_p is running **+1.18pp worse** than frieren #914 baseline.

### Validation Trajectory (run `6gd9u34e`)

| EP | abupt% | vol_p% | surf_p% | wall% | gate | result |
|---:|---:|---:|---:|---:|:---:|:---:|
| 1 | 17.8783 | 9.9910 | 12.9275 | 19.9261 | — | — |
| 2 | 11.8577 | 6.8048 |  8.4451 | 13.2260 | — | — |
| 3 | 10.4390 | 6.2625 |  7.3952 | 11.5707 | — | — |
| 4 |  9.7646 | 5.8770 |  6.9307 | 10.8180 | — | — |
| **5** | **9.0101** | **5.6907** | **5.9714** | **10.0299** | **≤7.5%** | **MISS (+1.51pp) → KILL** |

### Comparison vs frieren #914 baseline (`wdxtdmhy`)

| EP | mine abupt | frieren abupt | Δ abupt | mine vol_p | frieren vol_p | Δ vol_p |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 17.878 | 11.122 | +6.756 |  9.991 | 8.513 | +1.478 |
| 2 | 11.858 |  7.748 | +4.110 |  6.805 | 5.313 | +1.492 |
| 3 | 10.439 |  7.386 | +3.053 |  6.263 | 5.279 | +0.984 |
| 4 |  9.765 |  7.001 | +2.764 |  5.877 | 4.611 | +1.266 |
| 5 |  9.010 |  6.927 | +2.083 |  5.691 | 4.511 | +1.180 |

**Both metrics are persistently worse at every epoch.** The gap is shrinking on abupt (+6.76 → +2.08pp) but vol_p delta is roughly flat (+1.18 to +1.49pp), so vol_p is NOT receiving better signal — it's just being matched by the ongoing global descent.

### Loss-weight verification

Static weight is wired correctly in `train.py:337`: `weighted_volume_loss = volume_loss_weight * volume_loss`. With `--no-use-gradnorm`, `gradnorm_weights = None` (line 481-486 enabled only when `config.use_gradnorm`), so the 2.0 multiplier is applied unconditionally at every step and is **never adapted/equilibrated**. Confirmed via W&B config: `volume_loss_weight=2`, `use_gradnorm=False`. The static weight is not being cancelled — it is simply not helping.

### Train command

```bash
torchrun --nproc_per_node=8 train.py \
  --lr 1e-4 --weight-decay 0.005 --optimizer lion \
  --model-layers 5 --model-pe string_multisigma \
  --pe-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --no-use-gradnorm \
  --volume-loss-weight 2.0 \
  --use-y-symmetry-aug --y-symmetry-aug-prob 0.5 \
  --train-volume-points 65000 --eval-volume-points 65000 \
  --epochs 50 \
  --wandb-group vol-loss-weight-static \
  --wandb-name nezuko/vol-loss-weight-2.0-no-gradnorm \
  --agent nezuko
```

- **W&B run ID:** `6gd9u34e`
- **Peak GPU mem:** 17.72 GB (per-GPU, GPU1; range 16.0–17.7 across 8 ranks)
- **Killed:** 2026-05-10 ~02:01Z, partway into EP6 (~step 32k/274k total)
- **Wall clock:** ~67 min for 5 epochs + partial EP6

### What happened — hypothesis REJECTED

The hypothesis predicted that persistent 2× upweighting on `volume_loss` (without GradNorm cancellation) would force the model to prioritise vol_p and reduce the val→test gap. Three observations falsify this:

1. **vol_p got worse, not better.** At EP5, vol_p=5.69% vs frieren's 4.51% (+1.18pp). Even at EP4, vol_p=5.88% vs frieren's 4.61% (+1.27pp). The upweighting did not allocate more useful gradient toward vol_p — it allocated more *noisy* gradient.
2. **Other components got worse too.** Abupt is +2.08pp behind, surf_p is +1.45pp behind (5.97% vs ~4.52%), wall is +2.36pp behind. The 2× scaling raised the magnitude of the volume-pressure gradient but did not improve its information content; the optimizer treats it as larger noise.
3. **Convergence shape suggests scale-mismatch, not signal-mismatch.** EP1 was dramatic (+6.76pp behind), then the gap narrows steadily. This is consistent with the model spending early epochs absorbing the inflated vol_p gradient before the joint loss landscape stabilises — i.e., the loss surface is being *distorted* in a way the optimizer eventually compensates for, not *steered* toward better vol_p.

This is an interesting null on the "training-time signal" theory of the val→test vol_p gap. PR #911 showed GradNorm + static-2.0 self-cancels; this PR shows pure static-2.0 actively harms. Together they suggest **the val→test gap is NOT primarily a loss-weighting problem.** The "Neutral" outcome the advisor described ("vol_p unchanged → suggests vol_p gap is not a training-time signal problem") is closer to the truth than the "Failure" framing — except vol_p moved in the wrong direction, which is stronger evidence still.

### Suggested follow-ups

1. **Abandon static loss-weighting on vol_p.** Both PR #911 (with GradNorm) and PR #936 (without) failed; the upweight-vol_p direction should be considered closed. Save GPU-hours.
2. **Pivot to data-distribution alignment** (the advisor's stated next-step in the original PR body). Concrete options:
   - **OOD test-set diagnostics**: histogram-compare the test-set vol_p targets vs train+val. The +7pp val→test gap looks like a covariate-shift signature, not a training problem.
   - **Sample-reweighting or test-mimicking sampling** during training (e.g., reweight cases whose vol_p targets are closer to test-set quantiles).
   - **Test-set augmentation / TTA**: average predictions over y-flip + small crops at test time.
3. **If anyone retries vol_p weighting**, the only direction left is `<1.0` (e.g. 0.5) to *reduce* the noisy vol_p gradient and let the other components dominate — but this is just frieren's setup with a different multiplier on a quiet term, so likely a small effect.
4. **Lighter touch experiment**: try `volume-loss-weight 1.5` for fewer epochs (10–15) before committing 50 epochs — gate at EP5 ≤7.0% to detect the trend cheaply. (Not recommending we run this — listing only because the original PR mentioned it as a follow-up if vol_p had improved.)

### Why the early kill was correct

The advisor's gate at EP5 ≤7.5% was sensible: frieren hit 6.93% at EP5, so ≤7.5% allowed slack for stochastic variation. My run at 9.01% is a 1.51pp miss with a clear trajectory of *not* recovering — the EP4→EP5 drop was 0.75pp, and even an exceptional 2.5pp drop at EP6 only gets us to 6.5%, but the EP10 gate (7.2%) requires sustained improvement that cannot materialise when surf_p is also lagging. Continuing to EP30 would have cost ~7 more hours of GPU time on a confirmed null. Killing at EP5 saves the budget for the data-distribution pivot.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #936 because ADVISOR — EP5 Gate Miss: Hypothesis Falsified, Direction Closed

nezuko self-terminated correctly at EP5 with abupt=9.010% vs gate ≤7.5% (+1.51pp miss). The student's analysis is confirmed and thorough.

**EP1–EP5 Trajectory:**
| Epoch | abupt | vol_p |
|-------|-------|-------|
| EP1 | 17.878% | — |
| EP2 | 11.858% | 6.805% |
| EP3 | 10.439% | — |
| EP4 | 9.765% | — |
| EP5 | 9.010% | 5.691% |

**Hypothesis verdict: FALSIFIED**

vol-loss-weight=2.0 without GradNorm did NOT improve vol_p generalization. EP5 vol_p=5.691% vs frieren's 4.511% (+1.18pp WORSE). The static upweighting of vol loss actively harmed learning compared to the unweighted 5L+GradNorm baseline.

**The direction is now fully exhausted:**
- PR #911: static vol_loss_weight=2.0 WITH GradNorm → no-op (GradNorm self-cancels)
- PR #936: static vol_loss_weight=2.0 WITHOUT GradNorm → actively harmful

**Key insight confirmed:** The val→test vol_p gap (+7pp systematic) is NOT a training-time loss signal problem. Upweighting vol_p loss in any form does not close the gap. The gap is almost certainly a covariate shift / data distribution problem — the model overfits to the training distribution's vol_p patterns and fails to generalize OOD.

**What this means:** The remaining active hypothesis for vol_p improvement is PR #938 (fern) with 96k vol + 60k surface proportional sampling. This attacks data coverage, not loss weighting — a fundamentally different mechanism. If #938 also fails to close the gap, we will escalate to data distribution alignment and OOD regularization techniques.

Adding vol-loss-weight=2.0 (with or without GradNorm) to the confirmed dead-ends list..
